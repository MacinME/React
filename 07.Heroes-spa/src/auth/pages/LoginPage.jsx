import { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../context';

export const LoginPage = () => {

  const { setLogin } = useContext( AuthContext );

  const navigate = useNavigate();

  const onLogin = () => {

    setLogin('Max Shchew');

    const lastPath = localStorage.getItem('lastPath') || '/marvel'

    navigate(lastPath, {
      replace: true
    })
  }

  return (
    <div className="container mt-5">
      <h1>Login</h1>
      <hr />

      <button
        onClick={ onLogin }
        className="btn btn-primary"
      >
        Login
      </button>
    </div>
  )
}
