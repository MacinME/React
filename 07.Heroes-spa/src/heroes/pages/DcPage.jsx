import { HeroLists } from '../components';

export const DcPage = () => {
  return (
    <>
      <h1>DcPage</h1>

      <HeroLists publisher="DC Comics" />
    </>
  )
}
