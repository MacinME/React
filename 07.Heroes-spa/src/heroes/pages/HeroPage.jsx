import { Navigate, useNavigate, useParams } from 'react-router-dom'
import { getHeroByID } from '../helpers';
import { useMemo } from 'react';

export const HeroPage = () => {

  const { id } = useParams();
  
  const getHero = useMemo( () => getHeroByID( id ), [id]);
  const navigate = useNavigate();

  const onBack = () => {
    navigate(-1);
  }

  if(!getHero){
    return <Navigate to="/marvel" />
  }

  return (
    <div className="row mt-5 animate__animated animate__fadeInLeft">
        <div className="col-4">
          <img 
            src={`/assets/heroes/${ id }.jpg`} 
            alt={ getHero.superhero }
            className="img-thumbnail"  
          />
        </div>

        <div className='col-8'>
          <h3>{ getHero.superhero }</h3>
          <ul className='list-group -listgroup-flush'>
            <li className='list-group-item'><b>Alter Ego: </b> { getHero.alter_ego } </li>
            <li className='list-group-item'><b>Publisher: </b> { getHero.publisher } </li>
            <li className='list-group-item'><b>First Appearance: </b> { getHero.first_appearance } </li>
          </ul>

          <h5 className='mt-3'>Characters</h5>
          <p>
            { getHero.characters }
          </p>

          <button
            onClick={ onBack }
            className='btn btn-outline-primary'
          >
            Back
          </button>
        </div>
    </div>
  )
}
