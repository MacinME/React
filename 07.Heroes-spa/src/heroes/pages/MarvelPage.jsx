import { HeroLists } from '../components';

export const MarvelPage = () => {

  return (
    <>
      <h1>Marvel Page</h1>

      {
        <HeroLists publisher="Marvel Comics" />
      }
    </>
  )
}
