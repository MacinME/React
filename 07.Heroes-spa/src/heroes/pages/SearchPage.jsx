import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../helpers';
import { HeroCard } from '../components';

export const SearchPage = () => {

  const navigate = useNavigate();
  const location = useLocation();

  const { q = '' } = queryString.parse( location.search );
  const getHero = getHeroesByName( q );

  const {searchHero, onInputChange, onResetForm} = useForm({
    searchHero: q
  });

  const showSearch = (q.length <= 0);
  const showError  = (q.length > 0) && (getHero.length === 0);

  const onSubmit = (event) => {
    event.preventDefault();
    // if( searchHero.trim().length <= 1) return;
    navigate(`?q=${ searchHero }`);
  }

  return (
    <>
        <h1>Search</h1>
        <hr />

        <div className="row">
          <div className="col-5">
            <h4>Search Hero</h4>
            <hr />
            <form onSubmit={ onSubmit }>
              <input 
                type="text" 
                name="searchHero" 
                className="form-control" 
                placeholder="Search a hero"
                autoComplete="off"
                value={ searchHero }
                onChange={ onInputChange }
              />

              <button
                className="btn btn-outline-primary mt-1"
              >
                Search
              </button>
            </form>
          </div>

          <div className="col-7">
            <h4>Results</h4>
            <hr />
            <div 
              className="alert alert-primary"
              style={{ display: `${ showSearch ? '' : 'none' }` }}
            >
              search a hero
            </div>
            <div 
              className="alert alert-danger"
              style={{ display: `${ showError ? '' : 'none' }` }}
            >
              There's not hero <b>{ q }</b>
            </div>
            {
              getHero.map(hero => (
                <HeroCard key={ hero.id } {...hero} />
              ))
            }
          </div>
        </div>
    </>
  )
}
