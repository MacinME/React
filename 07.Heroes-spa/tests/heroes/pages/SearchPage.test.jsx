import { fireEvent, render, screen} from '@testing-library/react';
import { SearchPage } from '../../../src/heroes/pages/SearchPage';
import { MemoryRouter } from 'react-router-dom';

const mockedFunction = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedFunction
}))

describe('Test on <SearchPage />', () => { 

    
    test('should display the default state', () => { 

        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        )

        expect( container ).toMatchSnapshot();

    })

    test('should return the SpiderMan hero card', () => { 


        render(
            <MemoryRouter initialEntries={['/search?q=spider']}>
                <SearchPage />
            </MemoryRouter>
        )

        const input = screen.getByRole('textbox');
        expect( input.value ).toBe('spider');

        const img = screen.getByRole('img');
        expect( img.src ).toContain('/assets/heroes/marvel-spider.jpg');

        const alert = screen.getByLabelText('alert');
        expect( alert.style.display ).toContain("none");

    })

    test('should return error when hero does not exist', () => { 


        render(
            <MemoryRouter initialEntries={['/search?q=spiderman']}>
                <SearchPage />
            </MemoryRouter>
        )

        const alert = screen.getByLabelText('alert');
        expect( alert.style.display ).not.toBe("none");

    })

    test('should call the navigate custom hook when the form is hit', () => { 

        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        )

        const input = screen.getByRole('textbox');
        fireEvent.input( input, { target: { name: "searchHero", value: "spiderman" } } );

        const form = screen.getByLabelText('form');
        fireEvent.submit( form );
        expect( mockedFunction ).toHaveBeenCalledWith("?q=spiderman");


    })

})