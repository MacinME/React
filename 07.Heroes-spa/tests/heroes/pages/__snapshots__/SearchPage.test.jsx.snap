// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <SearchPage /> should display the default state 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Search Hero
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchHero"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary"
      >
        search a hero
      </div>
      <div
        aria-label="alert"
        class="alert alert-danger"
        style="display: none;"
      >
        There's not hero 
        <b />
      </div>
    </div>
  </div>
</div>
`;
